### PR TITLE
fix(object-hasher): switch to object-hash to fix hashing of large objects

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes https://github.com/pnpm/pnpm/pull/7563#issuecomment-1915260716

Marked this as a major bump given the hash values are changing.

Adds some more tests too; `object-hash` _cannot_ hash `undefined`, so I just made it return `0`. Feel free to suggest something better.

### CODE REVIEW

File: crypto/object-hasher/test/index.ts [hunk 1]
```diff
@@ -1,13 +1,25 @@
 import { hashObject, hashObjectWithoutSorting } from '@pnpm/crypto.object-hasher'
 
 describe('hashObject', () => {
+  const hash = hashObject
```
Comment (jakebailey) [inline crypto/object-hasher/test/index.ts:4]: This is just to make copy/pasting tests easier. I made mistakes while writing the tests until I realized I was testing the wrong function.

File: pnpm-lock.yaml [hunk 2]
```diff
@@ -18154,5 +18161,7 @@ packages:
     dev: false
 
 time:
+  /@types/object-hash@3.0.6: '2023-11-07T12:14:41.852Z'
```
Comment (jakebailey) [inline pnpm-lock.yaml:18164]: I... don't know why these are showing up in the lockfile?

**Review by zkochan — APPROVED** [review]

**Comment by zkochan** [comment]
:ship: 8.15.1

**Comment by jakebailey** [comment]
Thank you!

**Pull request merged**